src/zenml/logging/step_logging.py

```diff
@@ -473,28 +572,56 @@ def __init__(
         self.storage = PipelineLogsStorage(
             logs_uri=logs_uri, artifact_store=artifact_store
         )
+        self.artifact_store_handler: Optional[ArtifactStoreHandler] = None
         self.prepend_step_name = prepend_step_name
 
     def __enter__(self) -> "PipelineLogsStorageContext":
         """Enter condition of the context manager.
 
-        Wraps the `write` method of both stderr and stdout, so each incoming
-        message gets stored in the pipeline logs storage.
+        Creates and registers an ArtifactStoreHandler for log storage.
 
         Returns:
             self
         """
-        self.stdout_write = getattr(sys.stdout, "write")
-        self.stdout_flush = getattr(sys.stdout, "flush")
-
-        self.stderr_write = getattr(sys.stderr, "write")
-        self.stderr_flush = getattr(sys.stderr, "flush")
+        # Create storage handler
+        self.artifact_store_handler = ArtifactStoreHandler(self.storage)
 
-        setattr(sys.stdout, "write", self._wrap_write(self.stdout_write))
-        setattr(sys.stdout, "flush", self._wrap_flush(self.stdout_flush))
+        # Set formatter for storage handler
+        storage_formatter = logging.Formatter(
+            "[%(levelname)s] %(message)s (%(name)s:%(filename)s:%(lineno)d)"
+        )
+        self.artifact_store_handler.setFormatter(storage_formatter)
+
+        # Add handler to root logger
+        root_logger = logging.getLogger()
+        root_logger.addHandler(self.artifact_store_handler)
+
+        # Set root logger level to minimum of all active handlers
+        # This ensures records can reach any handler that needs them
+        self.original_root_level = root_logger.level
+        handler_levels = [handler.level for handler in root_logger.handlers]
+
+        # Set root logger to the minimum level among all handlers
+        min_level = min(handler_levels)
+        if min_level < root_logger.level:
+            root_logger.setLevel(min_level)
+
+        # Add handler to context variables for print() capture
+        handlers = logging_handlers.get().copy()
+        handlers.append(self.artifact_store_handler)
+        logging_handlers.set(handlers)
+
+        # Set the step names context variable
+        step_names_disabled = handle_bool_env_var(
+            ENV_ZENML_DISABLE_STEP_NAMES_IN_LOGS, default=False
+        )
 
-        setattr(sys.stderr, "write", self._wrap_write(self.stderr_write))
-        setattr(sys.stderr, "flush", self._wrap_flush(self.stderr_flush))
+        if step_names_disabled or not self.prepend_step_name:
+            # Step names are disabled through the env or they are disabled in the config
+            step_names_in_console.set(False)
```

Comment:
What if two get registered, the first one has prepend_step_name, the second one doesn't. The second one would incorrectly reset the value from the first one, no?

Reply:
Yes I agree that this will not happen, but I still think this is just bad/wrong code that could easily be fixed, and who knows where else this context will be used (maybe we allow users to create their own context to log into separate files for specific things?)